### PR TITLE
Add typescript definitions for built-in middlewares

### LINKED ISSUE #112: Add missing typescript definitions
I noticed this package doesn't have TypeScript definitions for the `middy/middlewares` submodule which makes it a little bit painful to use with TypeScript. Can I interest you in a pull request?

### PR DESCRIPTION
**Pull request by Juanelorganelo**
This pull requests adds typings for the middlewares public API. The file is in the root due to how typescript module resolution works for typings. For more information see [this thread](https://github.com/Microsoft/TypeScript/issues/8305)

Closes #112

### CODE REVIEW

**Comment by Juanelorganelo** [comment]
@lmammino I tried every possible way I could think of to keep the `middlewares.d.ts` file in the typescript folder. Apparently this haves to do with how [typescript resolves imports to submodules](https://github.com/Microsoft/TypeScript/issues/8305) so we can do `import * as middlewares from 'middy/middlewares'`. Thats why I left the file in the root and I didn't change the `typings` property of `package.json`. I could put a `postinstall` script to move the file to the root, but as metioned in the thread its kind of hacky and it doesn't work with `npm link`. Any feedback will be appreciated

**Comment by lmammino** [comment]
Sorry if this is taking too long, I have been very busy lately and I don't know enough about Typescript. It would be great to have some support by @joseSantacruz on this.

I'll try to play with this PR a bit and see if I can figure it out if we can avoid to keep the typescript file in the root folder

**Comment by lmammino** [comment]
@Juanelorganelo, I did some changes and my editor autocomplete seems to like the definitions now, but I am not sure I am providing types in the right way. Can you double check?

Also, it would be great to have some feedback from 2 typescript masters I know like @remojansen and @fsciuti 😇

I look forward to hearing from all of you guys, thanks in advance!

**Comment by joseSantacruz** [comment]
Hi @lmammino and @Juanelorganelo I put some comments on the PR, hope this help to improve this new feature

**Comment by remojansen** [comment]
It looks good but I would recommend adding a few test files under the `typescript` folder, like for example `types.test.ts`. This file should use the type definitions and then you can compile as them as part of the build. You can find a sample PR at https://github.com/godaddy/terminus/pull/4/files

File: typescript/middy.d.ts [hunk 1]
```diff
@@ -38,4 +38,4 @@ declare namespace middy {
   }
 }
 
-export = middy;
+export default middy;
```
[the remark this answers is not in the document]
Comment (remojansen) [reply typescript/middy.d.ts]: I had problems with this before, not sure if it is the same issues but the following helped me https://github.com/Microsoft/TypeScript/issues/5073

**Comment by Juanelorganelo** [comment]
Sorry for the delay, I've been really busy at work, anyway I just tested this with the remapped exports and `import * as middlewares from "middy/middlewares"` fails to compile with the `strict` flag. :'(

**Comment by lmammino** [comment]
Thanks @joseSantacruz and @remojansen! Your support is really really appreciated :)

@Juanelorganelo If it gets too tricky to have the typescript definitions inside the `typescript` folder I am ok for moving them back in the same directory of their related JS file.

**Comment by lmammino** [comment]
Thanks all for your valuable contribution on this!

@Juanelorganelo @joseSantacruz @remojansen 👍 🍻

**Pull request merged**